hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/FloatConverter.java

```diff
@@ -38,27 +40,66 @@ private FloatConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        float casted = cast(val);
+        // here the overflow may happen: (byte) casted = (byte) (int) casted
+        byte converted = (byte) casted;
+
+        // casts from float to int are saturating
+        if (converted != (int) casted || !Float.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.TINYINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public short asSmallint(Object val) {
-        return (short) cast(val);
+        float casted = cast(val);
+        // here the overflow may happen: (short) casted = (short) (int) casted
+        short converted = (short) casted;
+
+        // casts from float to int are saturating
+        if (converted != (int) casted || !Float.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.SMALLINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public int asInt(Object val) {
-        return (int) cast(val);
+        float casted = cast(val);
+        // casts from float to int are saturating
+        int converted = (int) casted;
+
+        // casts from float to long are saturating
+        if (converted != (long) casted || !Float.isFinite(casted)) {
+            throw numericOverflow(QueryDataTypeFamily.INT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public long asBigint(Object val) {
-        return (long) cast(val);
+        float casted = cast(val);
+        float truncated = (float) (casted > 0.0 ? Math.floor(casted) : Math.ceil(casted));
+        // casts from float to long are saturating
+        long converted = (long) truncated;
+
+        // No checks for NaNs and infinities are needed: NaNs are zeros and
```

Comment:
Why is this the case that NaN and infinities are handled silently? Consider not BIGINT, but DECIMAL. If we pass NaN or infinity to BigDecimal constructor, we will get an exception. I had to fix this in the math expressions branch. This seems reasonable - NaN is not zero, and infinity cannot be represented with a fixed value, because it is infinite.

Reply:
Please disregard